server/src/main/java/io/crate/expression/reference/DocRefResolver.java

```diff
@@ -116,34 +116,34 @@ public DocRefResolver(List<ColumnIdent> partitionedByColumns) {
                     } catch (ClassCastException | ConversionException e) {
                         // due to a bug: https://github.com/crate/crate/issues/13990
                         Object value = ValueExtractors.fromMap(response.getSource(), column);
-                        if (value instanceof List<?> list && list.stream().allMatch(Objects::isNull) &&
-                            !(ref.valueType() instanceof ArrayType<?>)) {
-                            return null;
-                        } else if (value instanceof Map<?, ?> map) {
-                            return replaceArraysWithNull(map, ref.valueType());
+                        try {
+                            return replaceArraysWithNull(value, ref.valueType());
+                        } catch (UnsupportedOperationException processException) {
+                            throw e;
                         }
-                        throw e;
                     }
                 });
         }
     }
 
     @SuppressWarnings({"unchecked", "rawtypes"})
-    private static Map<?, ?> replaceArraysWithNull(Map<?, ?> valueMap, DataType<?> valueType) {
-        Map newMap = new HashMap<>(valueMap.size());
-        for (var entry : valueMap.entrySet()) {
-            Object entryKey = entry.getKey();
-            Object entryValue = entry.getValue();
-            DataType<?> innerType = ((ObjectType) valueType).innerType((String) entryKey);
-            if (entryValue instanceof Map<?, ?> map) {
-                newMap.put(entryKey, replaceArraysWithNull(map, innerType));
-            } else if (entryValue instanceof List<?> list && list.stream().allMatch(Objects::isNull) &&
-                       innerType.id() != ArrayType.ID) {
-                newMap.put(entryKey, null);
-            } else {
-                newMap.put(entryKey, entryValue);
+    private static Object replaceArraysWithNull(Object value, DataType<?> valueType)
+        throws UnsupportedOperationException {
+
+        if (value instanceof List<?> list && list.stream().allMatch(Objects::isNull) &&
+            valueType.id() != ArrayType.ID) {
+            return null;
+        } else if (value instanceof Map<?, ?> valueMap) {
+            Map newMap = new HashMap<>(valueMap.size());
+            for (var entry : valueMap.entrySet()) {
+                Object entryKey = entry.getKey();
+                Object entryValue = entry.getValue();
+                DataType<?> innerType = ((ObjectType) valueType).innerType((String) entryKey);
+                newMap.put(entryKey, replaceArraysWithNull(entryValue, innerType));
             }
+            return newMap;
+        } else {
+            throw new UnsupportedOperationException("Different original error occured");
```

Comment:
Couldn't we return null here, and throw the original exception in the caller code instead if null? This would preserve the original exception and avoids another exception stack

Reply:
null is a valid return value of this method, so we need to distinguish the "error" branch.
I tried to pass the original Exception caught and re-throw it, if you think that's preferable, see new commit please.